consensus/src/transaction_shuffler/use_case_aware/delayed_queue.rs

```diff
@@ -0,0 +1,503 @@
+                        let new_use_case = entry.insert(UseCase::new_with_account(
+                            self.output_idx + 1,
+                            address,
+                            account,
+                        ));
+                        self.use_cases_by_delay
+                            .strict_insert(new_use_case.delay_key(), new_use_case_key.clone());
+                    },
+                }
+            }
+        }
+
+        Some(txn.txn)
+    }
+
+    /// Txn has to be delayed, attach it to respective account and use case.
+    fn queue_txn(
+        &mut self,
+        input_idx: InputIdx,
+        address: AccountAddress,
+        use_case_key: UseCaseKey,
+        txn: Txn,
+    ) {
+        match self.accounts.get_mut(&address) {
+            Some(account) => {
+                if account.is_empty() {
+                    // Account placeholder exists, move it from the placeholder queue to the main queue.
+                    self.account_placeholders_by_delay
+                        .remove(&account.delay_key());
+                    account.queue_txn(input_idx, txn);
+                    match self.use_cases.entry(use_case_key.clone()) {
+                        hash_map::Entry::Occupied(occupied) => {
+                            let use_case = occupied.into_mut();
+                            if use_case.is_empty() {
+                                self.use_case_placeholders_by_delay
+                                    .strict_remove(&use_case.delay_key());
+                            } else {
+                                self.use_cases_by_delay.strict_remove(&use_case.delay_key());
+                            }
+                            use_case.add_account(address, account);
+                            self.use_cases_by_delay
+                                .strict_insert(use_case.delay_key(), use_case_key.clone());
+                        },
+                        hash_map::Entry::Vacant(vacant) => {
+                            // Use case not tracked previously, the use case is ready at the current
+                            // output_idx, instead of output_idx +1 -- it makes a difference if
+                            // a txn later in the input queue that's of the same use case but not
+                            // blocked by account delay is tested for readiness.
+                            let use_case =
+                                UseCase::new_with_account(self.output_idx, address, account);
+                            self.use_cases_by_delay
+                                .strict_insert(use_case.delay_key(), use_case_key.clone());
+                            vacant.insert(use_case);
+                        },
+                    }
+                } else {
+                    // Account tracked and not empty, so appending a new txn to it won't affect positions
+                    // in delay queues
+                    account.queue_txn(input_idx, txn);
```

Comment:
why do we not need to touch use_case queues here?

Reply:
Because the account is not empty hence this txn won't be considered before the txns of the same account ahead of it.